execution/output_queue.go

```diff
@@ -2,80 +2,129 @@ package execution
 
 import (
 	"context"
+	"sync"
 
 	"github.com/golang/protobuf/proto"
-	"github.com/pkg/errors"
 
 	"github.com/cube2222/octosql/storage"
 )
 
+var outputQueues = sync.Map{}
+var outputQueuesFirstElement = sync.Map{}
+
 type OutputQueue struct {
-	tx storage.StateTransaction
+	tx            storage.StateTransaction
+	channel       chan proto.Message
+	firstElements chan proto.Message
 }
 
 func NewOutputQueue(tx storage.StateTransaction) *OutputQueue {
+	id := tx.WithPrefix(queueElementsPrefix).Prefix()
+	newQueueChan := make(chan proto.Message, 1024)
+	newFirstElementsChan := make(chan proto.Message, 1024)
+
+	channel, _ := outputQueues.LoadOrStore(id, newQueueChan)
+	firstElements, _ := outputQueues.LoadOrStore(id, newFirstElementsChan)
+
 	return &OutputQueue{
-		tx: tx,
+		channel:       channel.(chan proto.Message),
+		firstElements: firstElements.(chan proto.Message),
+		tx:            tx,
 	}
 }
 
 var queueElementsPrefix = []byte("$queue_elements$")
 
 func (q *OutputQueue) Push(ctx context.Context, element proto.Message) error {
-	queueElements := storage.NewDeque(q.tx.WithPrefix(queueElementsPrefix))
-
-	err := queueElements.PushBack(element)
-	if err != nil {
-		return errors.Wrap(err, "couldn't append element to queue")
-	}
+	q.channel <- element
 
 	return nil
 }
 
 func (q *OutputQueue) Peek(ctx context.Context, msg proto.Message) error {
-	queueElements := storage.NewDeque(q.tx.WithPrefix(queueElementsPrefix))
+	select {
+	case elem := <-q.firstElements:
+		q.firstElements <- elem
```

Comment:
how is this working? aren't we pushing `elem` to the END of `firstElements` yet we took it from the START of `firstElements`?